File: Remote/Source/Remote Elements/Commands/CommandSets/CommandSet.m
//
// CommandSet.m
// Remote
//
// Created by Jason Cardwell on 6/9/11.
// Copyright (c) 2011 Moondeer Studios. All rights reserved.
//

#import "CommandSet.h"
#import "CommandContainer_Private.h"
#import "Remote-Swift.h"

static int ddLogLevel   = DefaultDDLogLevel;
static int msLogContext = (LOG_CONTEXT_REMOTE | LOG_CONTEXT_FILE);

#pragma unused(ddLogLevel,msLogContext)

@interface CommandSet ()

@property (nonatomic, readwrite) CommandSetType type;
@property (nonatomic, strong)    NSSet        * commands;
@property (nonatomic, strong)    ButtonGroup  * buttonGroup;

@end

@interface CommandSet (CoreDataGeneratedAccessors)

- (void)addCommandsObject:(Command *)value;
- (void)removeCommandsObject:(Command *)value;
- (void)addCommands:(NSSet *)values;
- (void)removeCommands:(NSSet *)values;

@property (nonatomic, strong) NSMutableSet * primitiveCommands;
@property (nonatomic, strong) ButtonGroup  * primitiveButtonGroup;
@property (nonatomic, strong) NSNumber     * primitiveType;

@end


NSArray *sharedKeysForType(CommandSetType type) {
  static NSDictionary const * index = nil;
  static dispatch_once_t      onceToken;

  dispatch_once(&onceToken,
                ^{
    index = @{ @(CommandSetTypeDPad) :
               @[@(REButtonRoleDPadCenter),
                 @(REButtonRoleDPadUp),
                 @(REButtonRoleDPadDown),
                 @(REButtonRoleDPadLeft),
                 @(REButtonRoleDPadRight)],

               @(CommandSetTypeNumberpad) :
               @[@(REButtonRoleNumberpad1),
                 @(REButtonRoleNumberpad2),
                 @(REButtonRoleNumberpad3),
                 @(REButtonRoleNumberpad4),
                 @(REButtonRoleNumberpad5),
                 @(REButtonRoleNumberpad6),
                 @(REButtonRoleNumberpad7),
                 @(REButtonRoleNumberpad8),
                 @(REButtonRoleNumberpad9),
                 @(REButtonRoleNumberpad0),
                 @(REButtonRoleNumberpadAux1),
                 @(REButtonRoleNumberpadAux2)],

               @(CommandSetTypeRocker) :
               @[@(REButtonRoleRockerTop),
                 @(REButtonRoleRockerBottom)],

               @(CommandSetTypeTransport) :
               @[@(REButtonRoleTransportPlay),
                 @(REButtonRoleTransportPause),
                 @(REButtonRoleTransportStop),
                 @(REButtonRoleTransportRecord),
                 @(REButtonRoleTransportSkip),
                 @(REButtonRoleTransportReplay),
                 @(REButtonRoleTransportFF),
                 @(REButtonRoleTransportRewind)] };
  });

  return index[@(type)];
}

@implementation CommandSet

@dynamic buttonGroup, commands;

+ (instancetype)commandSetWithType:(CommandSetType)type inContext:(NSManagedObjectContext *)moc {
  CommandSet * commandSet = [self createInContext:moc];
  commandSet.type = type;
  return commandSet;
}

- (void)setObject:(Command *)command forKeyedSubscript:(NSString *)key {
  if (!command)
    ThrowInvalidNilArgument(command);

  else if (![self.index isValidKey:key])
    ThrowInvalidArgument(key, "is not valid for this command set");

  [self addCommandsObject:command];
  self.index[key] = command.permanentURI;
}

- (Command *)objectForKeyedSubscript:(id<NSCopying>)key {
  return ([self.index isValidKey:key]
          ? [Command objectForURI:self.index[key] context:self.managedObjectContext]
          : nil);
}

- (CommandSetType)type {
  [self willAccessValueForKey:@"type"];
  CommandSetType type = UnsignedIntegerValue(self.primitiveType);
  [self didAccessValueForKey:@"type"];
  return type;
}

/// Sets the type of the command set if the type specified represents a change, also creating a fresh index
/// @param type
- (void)setType:(CommandSetType)type {
  if (self.type != type) {
    [self willChangeValueForKey:@"type"];
    self.primitiveType = @(type);
    [self didChangeValueForKey:@"type"];

    if (CommandSetTypeIsValid(type)) {
      // ???: Should any existing index have its contents destroyed first?
      self.index = [MSDictionary dictionaryWithSharedKeys:sharedKeysForType(type)];
    }
  }
}

////////////////////////////////////////////////////////////////////////////////
#pragma mark Import and export
////////////////////////////////////////////////////////////////////////////////


- (void)updateWithData:(NSDictionary *)data {

  CommandSetType type = commandSetTypeFromImportKey(data[CommandSetTypeJSONKey]);

  if (CommandSetTypeIsValid(type)) {

    self.type = type;

    for (NSString * key in [data dictionaryByRemovingEntriesForKeys:@[CommandSetTypeJSONKey]]) {

      RERole role = remoteElementRoleFromImportKey(key);

      if (!(type & role)) continue; // Continue if type not composed into role

      NSDictionary * commandData = data[key];

      if (!isDictionaryKind(commandData)) continue; // Continue if commandData is not a valid dictionary object

      Command * command = [Command importObjectFromData:commandData context:self.managedObjectContext];

      if (command) self[@(role)] = command;

    }

  }

}

- (MSDictionary *)JSONDictionary {
  MSDictionary * dictionary = [super JSONDictionary];

  [dictionary removeObjectForKey:@"uuid"];

  dictionary[CommandSetTypeJSONKey] = CollectionSafe(commandSetTypeJSONValueForCommandSet(self));

  [self.index enumerateKeysAndObjectsUsingBlock:^(NSNumber * key, NSURL * uri, BOOL *stop) {
    SafeSetValueForKey([Command objectForURI:uri context:self.managedObjectContext].JSONDictionary,
                       roleJSONValueForRERole(UnsignedShortValue(key)),
                       dictionary);
  }];

  [dictionary compact];
  [dictionary compress];

  return dictionary;
}

@end

File: RemoteData/Model/RemoteElement/Commands/CommandSets/CommandSet.swift
//
//  CommandSet.swift
//  Remote
//
//  Created by Jason Cardwell on 3/2/15.
//  Copyright (c) 2015 Moondeer Studios. All rights reserved.
//

import Foundation
import CoreData
import MoonKit

@objc(CommandSet)
class CommandSet: CommandContainer {

  @NSManaged var commands: NSSet?

  static let sharedKeysByType: [CommandSetType:Set<RemoteElement.Role>] = [
    .Dpad: Set<RemoteElement.Role>([.Up, .Down, .Left, .Right, .Center]),
    .Transport: Set<RemoteElement.Role>([.Play, .Stop, .Pause, .Skip, .Replay, .FF, .Rewind, .Record]),
    .Numberpad: Set<RemoteElement.Role>([.One, .Two, .Three, .Four, .Five, .Six, .Seven, .Eight, .Nine, .Zero, .Aux1, .Aux2]),
    .Rocker: Set<RemoteElement.Role>([.Top, .Bottom])
  ]

  @objc enum CommandSetType: Int { case Unspecified, Dpad, Transport, Numberpad, Rocker }
  @NSManaged var primitiveType: NSNumber
  var type: CommandSetType {
    get {
      willAccessValueForKey("type")
      let type = primitiveType
      didAccessValueForKey("type")
      return CommandSetType(rawValue: type.integerValue) ?? .Unspecified
    }
    set {
      willChangeValueForKey("type")
      primitiveType = newValue.rawValue
      didChangeValueForKey("type")
      if let sharedKeys = CommandSet.sharedKeysByType[newValue] {
        index = MSDictionary(sharedKeys: Array(map(sharedKeys){$0.rawValue}))
      } else {
        index = MSDictionary()
      }
    }
  }

  /**
  subscript:

  :param: key RemoteElement.Role

  :returns: Command?
  */
  subscript(key: RemoteElement.Role) -> Command? {
    get { return index[key.rawValue] as? Command }
    set { index[key.rawValue] = newValue }
  }

  /**
  updateWithData:

  :param: data [String:AnyObject]
  */
  override func updateWithData(data: [String:AnyObject]) {
    super.updateWithData(data)

    if let moc = managedObjectContext,
      let typeJSON = data["type"] as? String {
      let type = CommandSetType(JSONValue: typeJSON)
      if type != .Unspecified {
        self.type = type
        let commands = data - "type"
        for (roleJSON, roleData) in commands {
          if let commandData = roleData as? [String:AnyObject],
            let command = Command.importObjectWithData(commandData, context: moc) {
              self[RemoteElement.Role(JSONValue: roleJSON)] = command
          }
        }
      }
    }
  }

  /**
  JSONDictionary

  :returns: MSDictionary!
  */
  override func JSONDictionary() -> MSDictionary {
    let dictionary = super.JSONDictionary()

    dictionary["type"] = type.JSONValue
    index.enumerateKeysAndObjectsUsingBlock { (key, uri, _) -> Void in
      if let command = self.managedObjectContext?.objectForURI(uri as! NSURL) as? Command {
        dictionary[RemoteElement.Role(rawValue: (key as! NSNumber).integerValue).JSONValue] = command.JSONDictionary()
      }
    }

    dictionary.compact()
    dictionary.compress()

    return dictionary
  }


}

extension CommandSet.CommandSetType: JSONValueConvertible {
  var JSONValue: String {
    switch self {
      case .Dpad:      return "dpad"
      case .Transport: return "transport"
      case .Numberpad: return "numberpad"
      case .Rocker:    return "rocker"
      default:         return "unspecified"
    }
  }

  init(JSONValue: String) {
    switch JSONValue {
      case "dpad":      self = .Dpad
      case "transport": self = .Transport
      case "numberpad": self = .Numberpad
      case "rocker":    self = .Rocker
      default:          self = .Unspecified
    }
  }
}